pkgs/fc/check-ceph/nautilus/fc/check_ceph/check_snapshot_restore.py

```diff
@@ -0,0 +1,251 @@
+            )
+    return exit_code
+
+
+def get_largest_snaps(connection: rados.Rados, pool_roots: dict) -> dict:
+    """Retrieves the largest snapshot per cluster root over all RBD pools specified for
+    that root in pool_roots.
+    The result is a dict of the form {cluster_name: (poolname, snapshotdata)}."""
+    largest_snaps = {}
+    for root, pools in pool_roots.items():
+        # Right now, we only need to deal with a single rbd pool per root, but let's
+        # keep this generic. As fill alerting happens at cluster root level, we need
+        # to find the largest rbd snapshot from all pools of that cluster.
+        root_largest_snap = root_dummy = ("dummypool", {"size_bytes": -1})
+        for pool in pools:
+            with contextlib.closing(connection.open_ioctx(pool)) as poolio:
+                pool_largest_snap = largest_snap_per_pool(poolio, pool)
+                if not pool_largest_snap:
+                    print(f"INFO: Pool {pool} has no (largest) snapshots.")
+                elif (
+                    pool_largest_snap["size_bytes"]
+                    > root_largest_snap[1]["size_bytes"]
+                ):
+                    root_largest_snap = (pool, pool_largest_snap)
+
+        if root_largest_snap is not root_dummy:
+            largest_snaps[root] = root_largest_snap
+    return largest_snaps
+
+
+def largest_snap_per_pool(poolio: rados.Ioctx, poolname: str):
+    largest_snap = dummy = {"size_bytes": -1}  # a pseudo-optional value
+    # in principle, only the ioctx determines the pool used. But for cleanliness always
+    # instantiate a new RBD per pool
+    rbdpool = rbd.RBD()
+    for imgname in rbdpool.list(poolio):
+        # we need an img object to query snapshots and their size
+        try:
+            snaps = rbd.Image(poolio, imgname).list_snaps()
+            for snap in snaps:
+                if snap["size"] > largest_snap["size_bytes"]:
+                    largest_snap = {
+                        "imgname": imgname,
+                        "snapname": snap["name"],
+                        "size_bytes": snap["size"],
+                    }
+        except rbd.ImageNotFound:
+            # as listing images and then accessing the images is not atomic, it is
+            # possible for the image to already be deleted. Ignore this, the changes
+            # will be considered at the next check run.
+            print(
+                f"INFO: Image {poolname}/{imgname} not found, possibly has been "
+                "deleted during the check run"
+            )
+            continue
+    return None if largest_snap is dummy else largest_snap
+
+
+def get_cluster_fillstats(connection, pool_roots):
+    """Parses the cluster root stats into a more digestible dictionary structure.
+    Only considers the pool roots of interest to us, specified in `pool_roots`."""
+    cluster_roots = (
+        node
+        for node in json.loads(
```

Comment:
i'd extract the json.loads() expression into a separate variable

Reply:
Or even better: make the output of the subprocess call the input into this (maybe expect the json.loads to have already been handled so you can provide test data more easily (and don't have to wrap them into multi-line json strings).

the same is true for any interaction with the cluster, actually: let `main()` have an early block that does all the outside-world interactions and then have the other functions operate on that data. this will result in a lot of mocking to become a lot easier in the tests.